kv/etcd/mock_test.go

```diff
@@ -0,0 +1,394 @@
+package etcd
+
+import (
+	"context"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+// Tests to make sure the mock Etcd client works which we need for kv.Client tests.
+// Quis custodiet ipsos custodes?
+
+func TestMockKv_Get(t *testing.T) {
+	t.Run("exact match", func(t *testing.T) {
+		pair := mvccpb.KeyValue{
+			Key:   []byte("/foo"),
+			Value: []byte("1"),
+		}
+
+		kv := newMockKV()
+		kv.values = map[string]mvccpb.KeyValue{string(pair.Key): pair}
+		res, err := kv.Get(context.Background(), "/foo")
+
+		require.NoError(t, err)
+		require.Len(t, res.Kvs, 1)
+		assert.Equal(t, []byte("/foo"), res.Kvs[0].Key)
+	})
+
+	t.Run("not exact match", func(t *testing.T) {
+		pair := mvccpb.KeyValue{
+			Key:   []byte("/foo"),
+			Value: []byte("1"),
+		}
+
+		kv := newMockKV()
+		kv.values = map[string]mvccpb.KeyValue{string(pair.Key): pair}
+		res, err := kv.Get(context.Background(), "/bar")
+
+		require.NoError(t, err)
+		assert.Empty(t, res.Kvs)
+	})
+
+	t.Run("prefix match", func(t *testing.T) {
+		pair1 := mvccpb.KeyValue{
+			Key:   []byte("/foo"),
+			Value: []byte("1"),
+		}
+		pair2 := mvccpb.KeyValue{
+			Key:   []byte("/baz"),
+			Value: []byte("2"),
+		}
+		pair3 := mvccpb.KeyValue{
+			Key:   []byte("/first"),
+			Value: []byte("3"),
+		}
+
+		kv := newMockKV()
+		kv.values = map[string]mvccpb.KeyValue{
+			string(pair1.Key): pair1,
+			string(pair2.Key): pair2,
+			string(pair3.Key): pair3,
+		}
+		res, err := kv.Get(context.Background(), "/f", clientv3.WithPrefix())
+
+		require.NoError(t, err)
+		assert.ElementsMatch(t, []*mvccpb.KeyValue{&pair1, &pair3}, res.Kvs)
```

Comment:
small nit, renaming these variables makes the assertion a little more immediately clear.

```suggestion
		fooPair := mvccpb.KeyValue{
			Key:   []byte("/foo"),
			Value: []byte("1"),
		}
		bazPair := mvccpb.KeyValue{
			Key:   []byte("/baz"),
			Value: []byte("2"),
		}
		firstPair := mvccpb.KeyValue{
			Key:   []byte("/first"),
			Value: []byte("3"),
		}

		kv := newMockKV()
		kv.values = map[string]mvccpb.KeyValue{
			string(pair1.Key): fooPair,
			string(pair2.Key): bazPair,
			string(pair3.Key): firstPair,
		}
		res, err := kv.Get(context.Background(), "/f", clientv3.WithPrefix())

		require.NoError(t, err)
		assert.ElementsMatch(t, []*mvccpb.KeyValue{&fooPair, &firstPair}, res.Kvs)
```